enhancements/machine-api/control-plane.md

```diff
@@ -0,0 +1,343 @@
+
+### Goals
+
+- To have a declarative mechanism to manage the Control Plane as a single entity.
+- To support declarative safe horizontal scaling towards an odd number of replicas for the control plane compute resources.
+- To support declarative safe vertical scaling for the control plane compute resources.
+- To support declarative safe self-healing and replacement of Control Plane compute resources when the healthy resources are above n/2.
+- To support even spread of compute resources across multiple failure domains.
+
+### Non-Goals / Future work
+
+- To integrate with any existing etcd topology e.g external clusters. Stacked etcd with dynamic member identities, local storage and the Cluster etcd Operator are an assumed invariant.
+- To managed individual Control Plane components. Self hosted Control Plane components that are self managed by their operators is an assumed invariant:
+	- Rolling OS upgrades and config changes at the software layer are managed by the Machine Config Operator.
+	- etcd Guard that ensures a PDB to honour quorum is managed by the Machine Config Operator.
+	- Each individual Control Plane component i.e Kube API Server, Scheduler and controller manager are self hosted and managed by their operators.
+	- The Cluster etcd Operator manages certificates, report healthiness and add etcd members as "Master" nodes join the cluster.
+	- The Kubelet is tied to the OS and therefore is managed by the Machine Config Operator.
+- To integrate with any Control Plane components topology e.g Remote hosted pod based Control Plane.
+- Automated disaster recovery of a cluster that has lost quorum.
+- To manage OS upgrades. This is managed by the Machine Config Operator.
+- To manage configuration changes at the software layer. This is managed by the Machine Config Operator.
+- To manage the life cycle of Control Plane components
+- To automate the provisioning and decommission of the bootstrapping instance managed by the installer.
+- To provide autoscaling support. This proposal is a necessary first step for autoscaling. It focuses on settling on the primitives that abstracts away the Control Plane as single scalable resource. In a follow up RFE we can discuss how/when to auto scale this resource based on relevant cluster metrics e.g number of workers, number of etcd objects, etc.
+
+## Proposal
+
+This proposes an ad-hoc CRD and controller for declaratively managing as a single entity the compute resources that host the OCP Control Plane components.
+
+This controller differs from a regular machineSet in that: 
+- It ensures that scaling operations are non disruptive for etcd:
+	- It scales one resource at a time.
+	- It let scaling operations proceed only when all etcd members are healthy and all the owned machines have a backed ready node.
```

Comment:
I think we can let `cluster-etcd-operator` conclude when scaling etcd is appropriate. Meaning if we have 3 node cluster and are going to replace node N. cluster-etcd-operator will watch for changes to your $resource and the scale up or down based on those observations. You can't roll the next node to upgrade until etcd comes back up as per PDB. We already watch Nodes so how will this improve our observability of change?